### PR TITLE
chore: move theorems to std4

### PR DESCRIPTION
**Pull request by chabulhwi**
* Move `String.csize_pos` to `Std.Data.Char`.
* Move `Nat.add_pos_left` and `Nat.add_pos_right` to `Std.Data.Nat.Lemmas`.
* Move lemmas on `List` to `Std.Data.List.Lemmas`.
* Move lemmas on `String` to `Std.Data.String.Lemmas`.
* Fix theorems and `test/matrix`.
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/String/Basic.lean [hunk 1]
```diff
@@ -159,26 +137,20 @@ theorem toList_singleton (c : Char) : (String.singleton c).toList = [c] :=
   rfl
 #align string.to_list_singleton String.toList_singleton
 
-theorem toList_nonempty : ∀ {s : String}, s ≠ "" → s.toList = s.head :: (s.popn 1).toList
```
Comment (chabulhwi) [inline Mathlib/Data/String/Basic.lean:162]: `String.popn` is a duplicate of `String.drop`. See https://github.com/leanprover-community/mathlib4/pull/3789.

**Comment by chabulhwi** [comment]
I don't know why the following test failed:
```shell
$ lake env lean test/matrix.lean
test/matrix.lean:16:0: warning: 'variables' has been replaced by 'variable' in lean 4
test/matrix.lean:54:15: error: unknown parser #guard_expr $r$eq:equal$p
test/matrix.lean:78:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:79:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:80:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:81:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:82:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:83:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:84:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:85:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
test/matrix.lean:86:0: error: elaboration function for 'Matrix.guardExprCmd' has not been implemented
```

**Comment by digama0** [comment]
`#guard_expr` was recently added to std, see leanprover/std4#133.

**Comment by kmill** [comment]
@chabulhwi Given that I moved these commands to std4, I pushed a change to fix `test/matrix.lean`

File: Mathlib/Data/List/Infix.lean [hunk 2]
```diff
@@ -43,58 +43,6 @@ variable {l l₁ l₂ l₃ : List α} {a b : α} {m n : ℕ}
 
 section Fix
 
-@[simp]
-theorem prefix_append (l₁ l₂ : List α) : l₁ <+: l₁ ++ l₂ :=
-  ⟨l₂, rfl⟩
-#align list.prefix_append List.prefix_append
```
Comment (digama0) [inline Mathlib/Data/List/Infix.lean:49]: You should keep the `#align` lines

**Comment by digama0** [comment]
This PR is redundant with #4012 . I suggest rebasing on top of it and/or closing this one

**Pull request closed without merging**